File: LeetCode/0377/cpp/lc0377.cpp
#include <cassert>
#include <vector>

class Solution {
public:
  int combinationSum4(std::vector<int>& nums, int target) {
    // In fact, it's a permutation sum, not a combination sum.

    std::vector<unsigned> dp;
    dp.resize(target + 1, 0);
    dp[0] = 1;
    for (int i = 1; i <= target; i++) {
      for (const auto& num: nums) {
        if (num <= i) {
          dp[i] += dp[i - num];
        }
      }
    }
    return dp[target];
  }
};

int main() {
  Solution s;

  std::vector<int> example1 = {1, 2, 3};
  assert(s.combinationSum4(example1, 4) == 7);

  std::vector<int> example2 = {9};
  assert(s.combinationSum4(example2, 3) == 0);
}

File: LeetCode/0377/golang/lc0377.go
package lc0377

// Note: it is in fact a permutation sum, not a combination sum.
// (in combinations the order of selected elements does not matter,
// but in this challenge the order matters, making it a permutation sum).
// https://en.wikipedia.org/wiki/Combination
func combinationSum4(nums []int, target int) int {
	dp := make([]int, target+1)
	dp[0] = 1 // a boundary condition: there's exactly one way to make a sum of 0 - by an empty set.
	for i := 1; i <= target; i++ {
		for _, num := range nums {
			if num <= i {
				dp[i] += dp[i-num]
			}
		}
	}
	return dp[target]
}
